File: linked-list/sll.cpp
#include <iostream>
#include <algorithm>

using namespace std;

struct Node
{
    int data;
    Node *prev;
    Node *next;
};

class SLinkedList
{
public:
    Node *head, *tail;

    SLinkedList()
    {
        head = NULL;
        tail = NULL;
    }

    void display();
    void displayReverse(Node *);
    void createNodes(int);
    void insertStart(int);
    void insertEnd(int);
    void insertPos(int, int);
    void deleteStart();
    void deleteEnd();
    void deletePos(int);
    void reverseList();
    void insertionSort();

    int getLength();
};

void SLinkedList::createNodes(int n)
{
    int data;
    for (int i = 0; i < n; i++)
    {
        cout << "Enter data for node #" << i + 1 << ": ";
        cin >> data;

        Node *new_node = new Node;
        new_node->data = data;

        if (head == NULL && tail == NULL)
        {
            head = new_node;
        }
        else if (head == tail)
        {
            head->next = new_node;
        }
        else
        {
            tail->next = new_node;
        }

        tail = new_node;
    }
}

void SLinkedList::display()
{
    if (head == NULL)
    {
        cout << "List empty.\n";
        return;
    }

    Node *temp = head;

    while (temp != NULL)
    {
        cout << temp->data << " ";
        temp = temp->next;
    }
}

void SLinkedList::displayReverse(Node* head)
{
    if (head == NULL)
    {
        return;
    }
    
    displayReverse(head->next);
    cout << head->data << " ";
}

void SLinkedList::insertStart(int data)
{
    Node *new_node = new Node;
    new_node->data = data;
    
    if (head == NULL)
    {
        tail = new_node;
    }
    else
    {
        new_node->next = head;
    }

    head = new_node;
}

void SLinkedList::insertEnd(int data)
{
    Node *new_node = new Node;
    new_node->data = data;

    if (head == NULL)
    {
        head = new_node;
    }
    else
    {
        tail->next = new_node;
    }
    
    tail = new_node;
}

void SLinkedList::insertPos(int data, int pos)
{
    Node *new_node = new Node;
    new_node->data = data;

    if (head == NULL || pos == 0)
    {
        insertStart(data);
    }
    else
    {
        Node *temp = head;
        
        while (pos--)
        {
            temp = temp->next;
        }

        new_node->next = temp->next;
        temp->next = new_node;
    }
}

int main()
{
    SLinkedList sll;

    sll.createNodes(4);
    
    sll.display();
    cout << endl;
    


    return 0;
}

File: linked-list/sll.py
import sys


try:
    sys.stdin = open(sys.path[0] + '/input.txt', 'r')
    sys.stdout = open(sys.path[0] + '/output.txt', 'w')
except FileNotFoundError:
    pass


class ListNode:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next


class SinglyLinkedList:
    def __init__(self):
        self.head = None

    def insert_start(self, val):
        if not self.head:
            self.head = ListNode(val)
        else:
            temp = ListNode(val)
            temp.next = self.head
            self.head = temp

    def insert_end(self, val):
        if not self.head:
            self.head = ListNode(val)
        else:
            temp = self.head
            while temp.next:
                temp = temp.next
            temp.next = ListNode(val)

    def insert_pos(self, val, pos):
        if not self.head or pos == 0:
            self.insert_start(val)
        else:
            pos -= 1
            temp = self.head
            for i in range(pos):
                if not temp.next:
                    return
                temp = temp.next

            new_node = ListNode(val)
            new_node.next = temp.next
            temp.next = new_node

    def delete_key(self, key):
        if self.head.val == key:
            self.head = self.head.next

        curr = self.head.next
        prev = self.head
        while curr:
            if curr.val == key:
                prev.next = curr.next
            prev = curr
            curr = curr.next

    def delete_pos(self, pos):
        if pos == 0:
            self.head = self.head.next

        prev = self.head
        curr = self.head.next

        for i in range(1, pos):
            prev = curr
            curr = curr.next

        if curr:
            prev.next = curr.next

    def get_length(self):
        length = 0
        curr = self.head
        while curr:
            curr = curr.next
            length += 1

        return length

    def get_length_recursive(self, node):
        if node:
            return 1 + self.get_length_recursive(node.next)
        return 0

    def display(self):
        temp = self.head
        while temp:
            print(temp.val, end=' ')
            temp = temp.next


sll = SinglyLinkedList()
